packages/apps/api/src/handlers/user.ts

```diff
@@ -31,6 +31,59 @@ export async function getUser(sierraClient: SierraClient, auth0Client: Auth0Clie
   }
 }
 
+export async function authenticateUser(auth0Client: Auth0Client, request: Request, response: Response) {
+  const email: string = request.body.email;
+  const password: string = request.body.password;
+
+  if (!isNonBlank(email) || !isNonBlank(password)) {
+    response.status(400).json(toMessage("All fields must be provided and non-blank"));
+  }
+
+  const validationResult = await auth0Client.validateUserCredentials(email, password);
```

Comment:
Should we not be checking for 404 here too, i.e. given email address not found? Does the Auth0 API return a specific error code that we can catch and communicate back to the caller if the account is blocked?

Reply:
Not in the status code, no. The OAuth2 specification does say that a human-readable error should be present in `error_description` on the response, but the status code will only ever be 200, 400, or 403.

Reply:
The docs do say that the message can be something like "MFA Auth required", but it can also show things like:

```json
{  "error_description":"Authorization pending: please repeat the request in a few seconds."  }
```

Reply:
Could we look the given email address up first, and if it isn't found, return a 404? Then if it does exist, proceed with the authentication request?

Reply:
Yeah, that should be fine to do in a private API.

Reply:
It does blow up this request by an extra 3 HTTP calls, which I think we should be conscious of. It might be better to change it later to get the error from whatever the token endpoint returns.

Reply:
What are the 3 calls? In my head it goes something like this:

1. Client calls API with email address and password
2. API calls Auth0 Managament API to validate email address exists
3. API calls Auth0 Authentication API to validate email address and password

If (2) or (3) fail, we return the appropriate status code back to the client immediately. To me checking that the email address exists and returning a `404` back to the caller if it doesn't adds 1 extra call.

Am I missing something?

Reply:
Ah it's 2 extra calls, not 3. I misread this bit:

```ts
    return axios.post(this.apiRoot + '/oauth/token', { // (1)
      client_id: this.clientId,
      client_secret: this.clientSecret,
      audience: this.apiAudience,
      grant_type: 'client_credentials'
    }, {
      validateStatus: status => status === 200
    }).then(response => {
      return axios.create({                           // misread as (2)
        baseURL: this.apiRoot + '/api/v2',
        headers: {
          Authorization: 'Bearer ' + response.data.access_token
        }
      });
    });
```

We weren't using a machine to machine token before, so we need to get that before we can validate the user exists.